### PR TITLE
Add/remove event listener hooks

### PR DESCRIPTION
**Pull request by matthewjh**
#142 

@vicb Please take a look, assuming travis tests pass.

@scamden Please also take a look. The public API for this is documented in README.md.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -242,7 +241,51 @@ growing the stack size on each recursive call.
...
+This hook allows you to intercept calls to `EventTarget#addEventListener`.
+
+````javascript
+var clickListenerCount = 0;
+
+zone.fork(
+  $addEventListener: function(parentAddEventListener) {
+    return function (type) {
+      if (type === 'click') clickListenerCount++;
+      return parentAddEventListener.call(this, arguments);
```
Comment (scamden) [inline README.md]: should this be apply?

File: README.md [hunk 2]
```diff
@@ -242,7 +241,51 @@ growing the stack size on each recursive call.
...
+This hook allows you to intercept calls to `EventTarget#removeEventListener`.
+
+````javascript
+var clickListenerCount = 0;
+
+zone.fork(
+  $removeEventListener: function(parentRemoveEventListener) {
+    return function (type) {
+      if (type === 'click') clickListenerCount--;
+      return parentRemoveEventListener.call(this, arguments);
```
Comment (scamden) [inline README.md]: should this be apply as well?

File: README.md (hunk 1 above)
Comment (matthewjh) [inline README.md]: It should. Thanks!

Comment (scamden) [inline README.md]: :)

**Comment by scamden** [comment]
this looks like it would solve my use case! i would still need to patch the handler, which feels slightly hairy but it might be over-engineering to provide an abstraction around overriding the callback

**Comment by matthewjh** [comment]
@scamden I take your point and can sympathise. This however seems like the simplest solution for now -- at least it provides max flexibility.

Another problem with overriding the handler/listener is that the `listener` argument can be an [EventListener](https://developer.mozilla.org/en-US/docs/Web/API/EventListener) object rather than a function. I think we can abstract that out easily so that the hook can always assume that the listener will be a function, otherwise the consumer will have to put in their own nasty logic around a `listener.apply` call.

File: lib/core.js [hunk 3]
```diff
@@ -119,7 +119,13 @@ Zone.prototype = {
   onZoneCreated: function () {},
   afterTask: function () {},
   enqueueTask: function () {},
-  dequeueTask: function () {}
+  dequeueTask: function () {},
+  addEventListener: function () {
+    return this['_addEventListener'].apply(this, arguments);
```
Comment (vicb) [inline lib/core.js]: `this._addEventListener` ?

**Comment by scamden** [comment]
ya i figured it wouldn't make sense to add, it's a bit overkill, just wanted to note it. this seems great!

File: lib/core.js (hunk 3 above)
Comment (matthewjh) [inline lib/core.js]: See https://github.com/angular/zone.js/pull/171/files#diff-50e3aa130a4f97a42ee2cf111c7b1d9dR92.

The `this` binding from the `addEventListener` call is used for the `apply` on `global.zone.addEventListener`. I don't see how else you could implement this as `patchEventTargetMethods` can be called on different objects when `EventTarget` isn't implemented.

Comment (vicb) [inline lib/core.js]: I only mean `this._add` instead of `this['_add']`

**Comment by vicb** [comment]
LGTM with one minor nit.

File: lib/core.js (hunk 3 above)
Comment (matthewjh) [inline lib/core.js]: Oh yeah -- not sure why I did that. Will correct.

**Comment by matthewjh** [comment]
@vicb Nit fixed, cheers.

**Comment by vicb** [comment]
landed as 18974403845a68dd450180b80c18fe20cb7a7959
Thanks !

**Comment by scamden** [comment]
thank you guys!

**Pull request closed without merging**